### snm3Cseq feature processing scripts

**About:**

This repository contains tools to process allc methylation files into genomic features. Computing the features' methylation fractions in both mCG and mCH, then dropping/imputing features based on applied coverage threshold parameters. 
### File Contents

**Scripts:** <br />
**allc_to_csv.py** : converts allc methylation file to four csvs (gene CH methylation, gene CG methylation, bin CH methylation, and bin CG methylation) for processing by allcDir_toScanpy.py. 
* Parameters:
	* allc_file_path : path to allc file on drive (.allc)
	* bin_size : base pair length of genomic bins to calculate across the entire genome
	* reference_genome_path : path to reference genome on drive (as .fasta or .fa)
	* genome_annotation_bed : bed_file of genomic annotations. (Annotations provided by the input bed are flanked +- 2kb)
	* out_prefix : prefix appended to output files
	
**allcDir_toScanpy.py** : Converts directory of csvs generated by allc_to_csv.py into AnnData objects with X matrix of methylation counts per feature (one for each modality, up to four; gmCH, gmCG, bmCH, and bmCG). Each AnnData object is saved as an .h5ad file along with an associated fastparquet (.parq) file containing coverage information.
* Parameters: 
	* path : path of directory on drive
	* outpre : prefix to append to output files

**frac_impute_sample.py** : Expects files from allcDir_toScanpy.py to be placed in a directory called sc_files. Computes feature methylation fractions, removes features with mean coverage < threshold, and imputes values by mean sample methylation for features of cells with coverage <= imputation threshold.
* Parameters : 
	* sample_root : input file prefixes leading _gmCH, gmCG, bmCH, or bmCG
	* min_mean_feature_coverage : remove features with mean coverage across all cells < this value
	* feature_impute_threshold : cells with coverage on features passing QC < this value will be imputed by the mean value for this feature

**Usage:**

```
python allc_to_csv.py $allc_file_path $bin_size $reference_genome_path $bed_file_path $out_prefix

python allcDir_toScanpy.py $allc_dir $out_prefix

python frac_impute_sample.py $sample_root
```
